src/ray/util/sequencer.h

```diff
@@ -0,0 +1,49 @@
+#ifndef RAY_UTIL_SEQUENCER_H_
+#define RAY_UTIL_SEQUENCER_H_
+
+#include <deque>
+#include <unordered_map>
+#include "absl/synchronization/mutex.h"
+
+namespace ray {
+
+/// \class Sequencer
+/// Sequencer guarantees that all operations with the same ordering key are sequenced.
+/// This class is thread safe.
+template <class KEY>
+class Sequencer {
+ public:
+  void execute_ordered(KEY key, std::function<void()> operation) {
+    mutex_.Lock();
+    pending_operations_[key].push_back(operation);
+    int queue_size = pending_operations_[key].size();
+    mutex_.Unlock();
+    if (1 == queue_size) {
+      operation();
+    }
+  }
+
+  void post_execute(KEY key) {
```

Comment:
What about not exposing this method?  We can pass a `done_callback` to `operation`. It will look like this:
```cpp
auto operation = [](SequencerDoneCallback done_callback) {
  // Do some time-consuming work here.
 done_callback(); // Inside `done_callback`, we'll call `post_execute`. But this is transparent to the users.
}
sequencer.execute_ordered(id, operation);
```

Also, I'd consider renaming `execute_ordered` to `post`. 

Any thoughts? @ffbin @ericl